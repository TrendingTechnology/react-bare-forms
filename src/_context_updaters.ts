// Context Updaters
// Methods to update & track the state of all inputs, validators & general form metadata
//
// For example:
// const context = {
//     // other props...
//     metadata: {
//          inputs: {
//              email: {
//                  isValid: false,
//                  messages: ['Cannot be empty'],
//                  value: '',
//                  isTouched: false,
//              },
//              username: {
//                  isValid: true,
//                  messages: [],
//                  value: 'joebloggs',
//                  isTouched: true,
//              }
//          }
//     }
// };
import {IValidation} from "./validators";


export const updateValidationMetadata = (context: any, update: any) =>
    (fieldName: string, fieldValue: any, validation: IValidation): void => {
        if(typeof context.metadata === "undefined") {
            return;
        }
        if(!(fieldName in context.metadata)) {
            update({
                ...context,
                metadata: {
                    ...context.metadata,
                    [fieldName]: {
                        ...validation,
                        value: fieldValue,
                        isTouched: true,
                    }
                },
            });
            // Should update context.metadata value
        } else if(context.metadata[fieldName].value != fieldValue) {
            update({
                ...context,
                metadata: {
                    ...context.metadata,
                    [fieldName]: {
                        ...validation,
                        value: fieldValue,
                        isTouched: true,
                    }
                },
            });
        }

    };


