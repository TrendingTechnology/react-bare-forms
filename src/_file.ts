import {default as React} from "react";
import {getFileFromRef} from "./uncrontrolled";

// File Helper Functions


/** @internal **/
export interface IFile {
    readonly name: string;
    readonly lastModified: string;
    readonly lastModifiedDate: string;
    readonly size: string;
    readonly type: string;
}

/** @internal */
export function createFileObject(ref: React.RefObject<HTMLFormElement>): IFile {
    try {
        let {name, lastModified, lastModifiedDate, size, type} = getFileFromRef(ref);
        return {
            name,
            lastModified,
            lastModifiedDate,
            size,
            type,
        };
    } catch (err) {
          return null;
    }
}
