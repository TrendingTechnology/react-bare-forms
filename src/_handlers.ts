import * as React from "react";

/** @internal */
type TypeHandler = (e: React.ChangeEvent<any>, name: string) => void;

/** @internal */
export function updateStateFromPassedInContext(parentState: any, setParentState: Function): TypeHandler {
    return (e: React.ChangeEvent<any>, name: string) => {
        setParentState({
            [name]: e.target.value,
        });
    };
}
