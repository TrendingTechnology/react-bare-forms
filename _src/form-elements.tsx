import * as React from "react";
import {FormElementValidators, FormContext, IFormContext} from "./form";
import {Validators} from "./validators";
import {selectTextField} from "./_helpers";
import {handleChange} from "./handlers";

export type FieldTypes =
    | "text"
    | "submit"
    | "textarea"
    | "select"
    | "password"
    | "email"
    | "checkbox";

interface ISubmitProps {
    children: any;
    type?: string;
}


interface IFormElementProps {
    name: any;
    label: string;
    type?: FieldTypes;
    hint?: string;
    validators?: Validators;
}

/**
 *
 * @param props
 * @constructor
 */
export const Submit = (props: ISubmitProps): any => {
    return <FormContext.Consumer>
        {({setIsSubmitted}: IFormContext) => {
            return <button
                disabled={false}
                type="submit"
                className="btn btn-primary"
                onClick={() => setIsSubmitted()}
            >{props.children}</button>
        }}
    </FormContext.Consumer>
};



/**
 *
 * @param props
 * @constructor
 */
export const Field = (props: IFormElementProps) => {
    const {name, label, type, hint = "", validators = null} = props;
    return(
        <FormContext.Consumer>
            {(context: IFormContext) => (
                <div className="form-group">
                    {label && <label>{label}</label>}
                    {hint && <small className="form-text text-muted">{hint}</small>}
                    {selectTextField(type, name, context)}
                    {props.validators && <FormElementValidators validators={validators} name={name} />}
                </div>
                )
            }
        </FormContext.Consumer>

    );
};

export const TextArea = (props: IFormElementProps) => {
    const {name, label, type, hint = "", validators = null} = props;
    return(
        <div className="form-group">
            {label && <label>{label}</label>}
            {hint && <small className="form-text text-muted">{hint}</small>}
            {/*<textarea className="form-control" rows={10}  onChange={handleChange(name)} />;*/}
            {props.validators && <FormElementValidators validators={validators} name={name} />}
        </div>
    );
};

export const Select = (props: IFormElementProps) => {
    const {name, label, type, hint = "", validators = null} = props;
    return(
        <div className="form-group">
            {label && <label>{label}</label>}
            {hint && <small className="form-text text-muted">{hint}</small>}
            {/*{selectTextField(type, name)}*/}
            {props.validators && <FormElementValidators validators={validators} name={name} />}
        </div>
    );
};




// return <div className="form-group">
//             //     <label>Instrument</label>
//             //     <select className="form-control">
//             //     {JSON.stringify(this.props.instruments)}
//             // {this.props.instruments && this.props.instruments.length ?
//             //     this.props.instruments.map((instr: IInstruments) =>
//             //         <option value={instr.instr_id}>{instr.name}</option>)
//             //     : null}
//             // </select>
//             // </div>;
