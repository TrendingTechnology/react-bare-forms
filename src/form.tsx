import {default as React, useEffect, useState} from "react";
import {updateStateFromPassedInContext} from "./_handlers";
import {IValidation} from "./validators";
import {updateRadioGroupMetadata, updateValidationMetadata} from "./_context_updaters";
import {IRadioField} from "./form-elements";


export interface IRadioGroupChildren {
    name: string;
    isChecked: boolean;
    disabled: boolean; // TODO
}

/** @internal */
export interface _IFormMetadata {
    messages: Array<string>;
    isValid: boolean;
    isTouched: boolean;
    value: any;
}

/** @internal */
export type TypeMetadata = { [k: string]: _IFormMetadata};
export type TypeRadioGroup = {[k: string]: IRadioGroupChildren};

export interface IMetdadata {
    fieldGroups: TypeRadioGroup;
    inputs: TypeMetadata;
}

/**
 * @interface **IForm** Exported Form interface available to the caller. Contains all the properties required by
 * the Form *RBF* Form's component.
 */
export interface IForm extends React.FormHTMLAttributes<HTMLFormElement> {
    /** The passed in state from the parent component */
    state: any;
    /** If the parent component is a class component then the context must contain the parent's **this** keyword. */
    context?: any;
    /** If the form's state is not directly contained at parent's state root object, then formKey needs to represent the key. */
    formKey?: string;
    /** Default is false. If set to true then the form state will be displayed just below the form component in `code` tags. */
    debug?: boolean;
    /** Default is True. If set to false, the  validators will not be called until the form is submitted. */
    dynamic?: boolean;
    /** The passed through props to the Form component */
    children?: any;
    /** Default set to false. If set to true then it will now use bootstrap styling or any extra elements. */
    bare?: boolean;
}

/** @internal */
export interface IFormContext {
    bare?: boolean;
    debug?: boolean;
    dynamic?: boolean;
    formKey?: string;
    metadata: TypeMetadata;
    state: any;
    updateParentState?: (e: React.ChangeEvent<any>, name: string) => void;
    updateFieldValidation?: (fieldName: string, fieldValue: any, validation: IValidation) => void;
    updateFormGroupMetadata?: (fieldGroupKey: string, radioProps: Array<{ props: IRadioField}>) => void;
}


const providerContext: IFormContext = {
    bare: false,
    state: {},
    formKey: null,
    debug: false,
    dynamic: true,
    metadata: { inputs: ({} as any), fieldGroups: ({} as any)},
};

/** @internal */
export const FormContext = React.createContext(providerContext);
/** @internal */
export const FormProvider = FormContext.Provider;
/**
 * @var Access the form state within a React Conext Api Consumer scope
 * @example For example:
 * ```
 * <Form state={this.state} context={this}>
 *      <FormConsumer>
 *      {({state}) => {
 *          return <div>{state} is equal {myState}</div>;
 *      }}
 *    </FormConsumer>
 * </Form>
 * ```
 * */
export const FormConsumer = FormContext.Consumer;


export const handleSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault();
};

/**
 *
 * @param props
 * @constructor
 */
export const Submit = (props: any) => {
    return <button type="submit">Submit</button>;
};


/**
 * The main Form component
 * @param props
 *
 * ```
 * // Minimal setup for a RBF's Form component
 *
 *  const myState = {
 *      username: '',
 *  }
 *
 *  <Form state={myState}></Form>
 * ```
 * @constructor
 */
export const Form = (props: IForm) => {
    const [parentState, setParentState] = useState(props.state);

    // If the parent component is a class component, then the state needs to be updated from the parent context
    if(props.context) {
        useEffect(() => {
            props.context.setState({
                ...parentState,
            });
        }, [parentState]);
    }


    const [context, updateContext] = useState(providerContext);

    const _providerContext: IFormContext = {
        bare: props.bare || context.bare,
        state: props.state,
        formKey: props.formKey,
        debug: props.debug || context.debug,
        dynamic: props.dynamic || context.dynamic,
        metadata: context.metadata,
        updateParentState: updateStateFromPassedInContext(parentState, setParentState),
        updateFieldValidation: updateValidationMetadata(context, updateContext),
        updateFormGroupMetadata: updateRadioGroupMetadata(context, updateContext),
    };

    return (
        <FormProvider value={_providerContext}>
            <form onSubmit={handleSubmit} {...props}>{props.children}</form>
        </FormProvider>
    );
};
