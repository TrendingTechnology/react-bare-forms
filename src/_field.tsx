import {FormContext, IFormContext} from "./form";
import {FormElementValidators, mergeDefaultCssWithProps} from "./_helpers";
import {ChangeEvent, default as React, ReactElement, useContext, useEffect} from "react";
import {
    IFileField,
    RadioGroupContext,
    TypeSelectCssSizeName
} from "./elements";
import {shouldUpdateRadioGroupContext} from "./_context_updaters";

interface IFieldClass<T> {
    create: (context: IFormContext) => ReactElement<T>;
    type: string;
    props: T;
}

/** @internal */
function _genericFormGroup<T extends any>(props: T, children: any) {
    return (
        <div className="form-group">
            {props.labelText && <label>{props.labelText}</label>}
            {children}
            {props.hint && <small className="form-text text-muted">{props.hint}</small>}
        </div>
    );
}

/** @internal */
abstract class _field<PropsType extends any> {

    type: string;

    props: PropsType;

    protected constructor(props: PropsType, type?: string) {
        this.type = type;
        this.props = props;
    }

    public createField(fieldCallback: Function) {
        const context: IFormContext = useContext(FormContext);
        const _validate = this.props.validators ?
            <FormElementValidators
                validators={this.props.validators}
                name={this.props.name}
                value={this.props.value}
                type="inputs"
            /> :
            null;
        if(context.bare) {
            return (<>{fieldCallback(context)}{_validate}</>);
        } else {
            return(<>{this.formGroup(fieldCallback(context))}{ _validate}</>);
        }
    }

    public abstract formGroup(children: any): ReactElement;

    public abstract getField(): (context: IFormContext) => ReactElement;

    static mergeDefaultCssWithProps(defaultValue: string, cssProps: any, bare: boolean): string {
        let cssStr = "";
        if(!bare) {
            cssStr += `${defaultValue} `;
        }
        if (cssProps) {
            cssStr += `${cssProps}`;
        }
        return cssStr;
    }

    public overrideEvent(e: any, value: any) {
        return {
            ...e,
            target: {
                ...e.target,
                value: !value,
            }
        }
    }

}

/** @internal */
export class InputField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(type: string, props: T) {
        super(props, type);
        this.type = type;
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        return (context: IFormContext) => {
            return <input
                type={this.type}
                value={context.state[this.props.name] || ""}
                onChange={(e) => context.updateParentState(e, this.props.name)}
                name={this.props.name}
                className={_field.mergeDefaultCssWithProps("form-control", this.props.className, context.bare)}
            />;
        }
    }

}

/** @internal */
export class CheckBoxField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(type: string, props: T) {
        super(props, type);
        this.type = type;
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return (
            <div className="form-group form-check">
                {children}
                {this.props.labelText && <label className="form-check-label">{this.props.labelText}</label>}
                {this.props.hint && <small className="form-text text-muted">{this.props.hint}</small>}
            </div>
        );
    }

    public getField() {
        return (context: IFormContext) => {
            return <input
                type={this.type}
                checked={context.state[this.props.name] || false}
                onChange={(e) => context.updateParentState(this.overrideEvent(e, context.state[this.props.name]), this.props.name)}
                name={this.props.name}
                className={_field.mergeDefaultCssWithProps("form-check-input", this.props.className, context.bare)}
            />;
        }
    }
}

/** @internal */
export class TextAreaField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(props: T) {
        super(props);
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        const {rows = 5} = this.props;
        return (context: IFormContext) => {
            return (
                <textarea
                    className={mergeDefaultCssWithProps("form-control", this.props.className, context.bare)}
                    rows={rows}
                    value={context.state[this.props.name] || ""}
                    onChange={(e) => context.updateParentState(e, this.props.name)}
                    name={this.props.name}
                />
            );
        }
    }
}

/** @internal */
export class RadioField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(type: string, props: T) {
        super(props, type);
        this.type = type;
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return (
            <div className="form-group form-check">
                {children}
                {this.props.labelText && <label className="form-check-label">{this.props.labelText}</label>}
                {this.props.hint && <small className="form-text text-muted">{this.props.hint}</small>}
            </div>
        );
    }

    public getField() {
        const radioContext: any = useContext(RadioGroupContext); // TODO - any
        return (context: IFormContext) => {
        const updateContexts = (e: React.ChangeEvent<any>, context: IFormContext) => {
            context.updateRadioGroupStateFromPassedInContext(
                this.overrideEvent(e, context.state[this.props.name]),
                this.props.name,
                radioGroup,
            );
            if(shouldUpdateRadioGroupContext(radioContext.children, context, radioContext.parent.name)) {
                useEffect(() => {
                    context.updateRadioGroupMetadata(radioContext.parent.name, radioContext.children);
                }, [radioContext]);

            }
        };
        const radioGroup = context.metadata.fieldGroups[radioContext.parent.name];
        return <input
            type={this.type}
            checked={context.state[this.props.name] || false}
            name={this.props.name}
            className={_field.mergeDefaultCssWithProps("form-check-input", this.props.className, context.bare)}
        />

        }
    }

}

/** @internal */
export class SelectField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(props: T) {
        super(props);
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        const {options = [], size = "default"} = this.props;
        return (context: IFormContext) => {

            return (
                <select
                    onChange={(e) => context.updateParentState(e, this.props.name)}
                    name={this.props.name}
                    className={_field.mergeDefaultCssWithProps(this.getSelectCssName(this.props.size), this.props.className, context.bare)}
                >
                    {options.map((optVal: string, i: number) => {
                        return <option value={optVal} key={i}>{optVal}</option>
                    })}
                </select>
            );
        }
    }

    private getSelectCssName(name: TypeSelectCssSizeName) {
        if(name === "default") {
            return "form-control"
        } else {
            return `form-control form-control-${name}`;
        }
    }
}

/** @internal */
export class FileField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(props: T) {
        super(props);
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        return (context: IFormContext) =>
            (<input
                ref={this.props.ref}
                type="file"
                className={_field.mergeDefaultCssWithProps("form-control-file", this.props.className, context.bare)}
            />);
    }
}
