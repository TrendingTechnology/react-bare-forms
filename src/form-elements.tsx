import {default as React, ReactElement} from "react";
import {FormConsumer, IFormContext} from "./form";
import {IValidators} from "./validators";
import {FormElementValidators, mergeDefaultCssWithProps} from "./_helpers";



export interface IField {
    /** The name of the form element (this should match the state property that you want be updated by this form element) */
    name: string;
    /** The state property that gets updated by this input field */
    value: any;
    /** If *bare* (see {@link IForm.bare}) is set to true then *labelText* will be inserted within **label** tags. */
    labelText?: string;
    /** If *bare* (see {@link Form.bare}) is set to true then *hint* will be inserted within **small** tags. */
    hint?: string;
    /** These are the {@link IValidators} that you can pass in the validate the form element. */
    validators?: IValidators;
    /** Add additional css */
    className?: string;
}

interface ITextInputField extends IField {}

interface IPasswordField extends IField {}

interface IEmailField extends IField {}

interface ICheckBoxField extends IField {}

interface ITextAreaField extends IField {
    rows?: number;
}

/**
 *
 * @param props
 * ```
 * // A bare form example ... remember to set the {@link Form.bare} property to `true`
 * <TextInputField
 *    value={this.state.username}
 *    name="username"
 * />
 *
 * // Example with Bootstrap styling (Bootstrap styling comes as default)
 *
 * <TextInputField
 *    value={this.state.username}
 *    name="username"
 *    hint="Needs to be at least 50 characters long"
 *    labelText="Username"
 *  />
 * ```
 * @constructor
 */
export const TextInputField = (props: ITextInputField) =>
    _createTextInputField("text")(props);

/**
 *
 * @param props
 * ```
 * // A bare form example ... remember to set the {@link Form.bare} property to `true`
 * <PasswordField
 *    value={this.state.username}
 *    name="username"
 *    validators={[isEmailValid()]}
 * />
 *
 * // Example with Bootstrap styling (Bootstrap styling comes as default)
 *
 * <PasswordField
 *    value={this.state.password}
 *    name="password"
 *    hint="Needs to be at least 8 characters long"
 *    labelText="Password"
 *    validators={[isEmailValid()]}
 *  />
 *  ```
 * Also we can create two *PasswordField* components to confirm passwords are equal. Please see
 * {@link areFieldsEqual} for more info.
 * The first *PasswordField* has has a *name* prop of **password** & the second *PasswordField* a name
 * prop of *confirmPassword*. Then we can add a {@link areFieldsEqual} validator to the *PasswordField*
 * with the *confirmPassword* name props (also notice how {@link areFieldsEqual} takes the first *PasswordField*
 * name as an argument).
 *
 * ```
 * <PasswordField
 *  name="password"
 *  // other props...
 *
 *  validators={[isEmailValid()]}
 * />
 *
 * <PasswordField
 *  name="confirmPassword"
 *  // other props...
 *  validators={[areFieldsEqual("password")]}
 * />
 * ```
 *
 * @constructor
 */
export const PasswordField = (props: IPasswordField) =>
    _createTextInputField("password")(props);

/**
 *
 * @param props
 * ```
 * // A bare form example ... remember to set the {@link Form.bare} property to `true`
 * <TextInputField
 *    value={this.state.username}
 *    name="username"
 * />
 *
 * // Example with Bootstrap styling (Bootstrap styling comes as default)
 *
 * <TextInputField
 *    value={this.state.username}
 *    name="username"
 *    hint="Needs to be at least 50 characters long"
 *    labelText="Username"
 *  />
 * ```
 * @constructor
 */
export const EmailField = (props: IEmailField) =>
    _createTextInputField("email")(props);


/**
 *
 * @param props
 * @constructor
 */
export const CheckBoxField = (props: ICheckBoxField) =>
    _createTextInputField("checkbox")(props);


/**
 * @internal
 * @param type
 * @private
 */
function _createTextInputField(type: string) {
    return (props: ITextInputField) => (
        <FormConsumer>
            {(context: IFormContext) => {
                const _input = <input
                    type={type}
                    value={(context as any)[props.name]}
                    onChange={(e) => context.updateParentState(e, props.name)}
                    name={props.name}
                    className={mergeDefaultCssWithProps("form-control", props.className, context.bare)}
                />;
                const _validate = props.validators ? <FormElementValidators validators={props.validators} name={props.name} />: null;
                if(context.bare) {
                    return (
                        <>
                            {_input}
                            {_validate}
                        </>
                    );
                } else {
                    return (
                        <FormGroup labelText={props.labelText} hint={props.hint}>
                            {_input}
                            {_validate}
                        </FormGroup>
                    )
                }

            }}
        </FormConsumer>
    );
}


/**
 * The TextAreaField takes in an extra prop of *row* which is a number & declares
 * the number of rows displayed by the textarea element. The TextAreaField accepts
 * all the {@link IField} props.
 * ```
 * // A bare form example ... remember to set the {@link Form.bare} property to `true`
 * <TextAreaField
 *    value={this.state.about}
 *    name="about"
 *    validators={[isFieldEmpty(20)]}
 * />
 *
 * // Example with Bootstrap styling (Bootstrap styling comes as default)
 *
 * <TextAreaField
 *    name="about"
 *    value={this.state.about}
 *    hint="Must be at least 20 characters"
 *    labelText="About you..."
 *    validators={[isFieldEmpty(20)]}
 * />
 * ```
 * @param props
 * @constructor
 */
export const TextAreaField = (props: ITextAreaField) => {
    const { rows = 5 } = props;

    return (
        <FormConsumer>
            {(context: IFormContext) => {
                let _textArea: ReactElement<ITextAreaField> =
                    <textarea
                        className={mergeDefaultCssWithProps("form-control", props.className, context.bare)}
                        rows={rows}
                        value={(context as any)[props.name] }
                        onChange={(e) => context.updateParentState(e, props.name)}
                        name={props.name}
                    />;
                const _validate = props.validators ? <FormElementValidators validators={props.validators} name={props.name} />: null;
                if(context.bare) {
                    return <>
                        {_textArea}
                        {_validate}
                    </>;
                } else {
                    return (
                        <div className="form-group">
                            <label>{props.labelText}</label>
                            {_textArea}
                            {_validate}
                        </div>
                    )
                }
            }}
        </FormConsumer>
    );
};


/** @internal */
interface IFormGroup {
    children: any;
    labelText?: string;
    hint?: string;
}

/**
 * @internal
 * @param props
 * @constructor
 */
export function FormGroup(props: IFormGroup): React.ReactElement {
    return (
        <div className="form-group">
            {props.labelText && <label>{props.labelText}</label>}
            {props.children}
            {props.hint && <small className="form-text text-muted">{props.hint}</small>}
        </div>
    );
}
