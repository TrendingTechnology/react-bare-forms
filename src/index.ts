import {Form, FormConsumer} from "./form";
import {
    CheckBoxField,
    EmailField, FileField,
    PasswordField, RadioField, RadioGroup,
    SelectField,
    SubmitButton,
    TextAreaField,
    TextInputField
} from "./elements";
import {areFieldsEqual, isChecked, isEmailValid, isFieldEmpty, isFile, isRadioChecked} from "./validators";
import {createFileRef, getFileFromRef} from "./uncrontrolled";




export {
    // Uncontrolled
    createFileRef,
    getFileFromRef,
    // Validators
    areFieldsEqual,
    isChecked,
    isEmailValid,
    isFieldEmpty,
    isFile,
    isRadioChecked,
    // Form Components
    FormConsumer,
    Form,
    SubmitButton,
    TextInputField,
    EmailField,
    PasswordField,
    TextAreaField,
    CheckBoxField,
    SelectField,
    FileField,
    RadioField,
    RadioGroup,
}
