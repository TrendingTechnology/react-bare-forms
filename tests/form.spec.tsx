import * as React from "react";
import TestRenderer from "react-test-renderer";
import {Form} from "../src/form";
import {
    act,
    findRenderedDOMComponentWithClass,
    renderIntoDocument,
} from "react-dom/test-utils";
import {unmountComponentAtNode } from "react-dom";
import {TextInputField} from "../src/form-elements";
import { shallow, render } from 'enzyme';


let container: any = null;

beforeEach(() => {
   container = document.createElement("div");
   document.body.appendChild(container);
});


afterEach(() => {
   unmountComponentAtNode(container);
   container.remove();
   container = null;
});

describe("<Form />", () => {

    it("should render form tags with no other elements", () => {

        act(() => {
            render(
                <Form state={{}} bare={true}>

                </Form>, container);
        });
         expect(container.querySelector("form").childElementCount).toBe(0);
    });

    it("should contain a context object with state and form props", () => {
        let state = {
            username: "joebloggs"
        };
        let context = {state};
        let testFormRederer: any = TestRenderer.create(
            <Form state={state} ></Form>
        );

        const testFormInstance = testFormRederer.root;
        expect(testFormInstance.props).toEqual(context);
    });

    it("should render a form element with no extra tags", () => {
        let state = {
            username: "joebloggs"
        };

        const tree = TestRenderer.create(
            <Form state={state}>
                <TextInputField
                    name="username"
                    value={state.username}
                />
            </Form>
        ).toJSON();
        expect(tree).toMatchSnapshot();

        const wrapper = render(
            <Form state={state} bare={true}>
            <TextInputField
                name="username"
                value={state.username}
            />
        </Form>);

        expect(wrapper.find(".form-group")).toHaveLength(0);
        expect(wrapper.find(".form-control")).toHaveLength(0);
    });

    it("should render a form element with bootstrap extra tags", () => {
        let state = {
            username: "joebloggs"
        };

        const tree: any = TestRenderer.create(
            <Form state={ state}>
                <TextInputField
                    name="username"
                    value={state.username}
                />
            </Form>
        );
        expect(tree.toJSON()).toMatchSnapshot();

        const wrapper = render((<Form state={state}>
            <TextInputField
                name="username"
                value={state.username}
            />
        </Form>));

        expect(wrapper.find(".form-group")).toHaveLength(1);
        expect(wrapper.find(".form-control")).toHaveLength(1);
    });
});
