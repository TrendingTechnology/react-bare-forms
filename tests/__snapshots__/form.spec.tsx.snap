// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render a form element with bootstrap extra tags 1`] = `
<form
  onSubmit={[Function]}
  state={
    Object {
      "username": "joebloggs",
    }
  }
>
  <div
    className="form-group"
  >
    <input
      className="form-control "
      name="username"
      onChange={[Function]}
      type="text"
    />
  </div>
</form>
`;

exports[`<Form /> should render a form element with no extra tags 1`] = `
<form
  onSubmit={[Function]}
  state={
    Object {
      "username": "joebloggs",
    }
  }
>
  <div
    className="form-group"
  >
    <input
      className="form-control "
      name="username"
      onChange={[Function]}
      type="text"
    />
  </div>
</form>
`;
